src/forms/constraints/numbers.rs

```diff
@@ -0,0 +1,147 @@
+use crate::forms::FormValue;
+
+use super::FormConstraint;
+
+/// Validates that the input is an integer (positive or negative)
+pub struct Integer {
+    message: String,
+}
+
+impl Integer {
+    pub fn new(message: &str) -> Box<Self> {
+        Box::new(Self {
+            message: String::from(message),
+        })
+    }
+}
+
+impl FormConstraint for Integer {
+    fn validate(&self, output: &FormValue) -> bool {
+        match output {
+            FormValue::Nothing => true,
+            FormValue::Boolean(_) => true,
+            FormValue::Index(_) => true,
+            FormValue::String(value) => {
+                if let Some(chr) = value.chars().next() {
+                    if !chr.is_digit(10) && chr != '-' && chr != '+' {
+                        return false;
+                    }
+                }
+                value.chars().skip(1).all(|x| x.is_digit(10))
+            }
+            FormValue::Map(entries) => entries.iter().all(|(_, x)| self.validate(x)),
+            FormValue::List(entries) => entries
+                .iter()
+                .all(|x| self.validate(&FormValue::String(String::from(x)))),
+            FormValue::Vec(entries) => entries.iter().all(|x| self.validate(x)),
+        }
+    }
+
+    fn get_message(&self) -> &str {
+        &self.message
+    }
+}
+
+/// Validates that the input is a number, allowing one comma or dot if it's a decimal number
+pub struct Number {
+    message: String,
+}
+
+impl Number {
+    pub fn new(message: &str) -> Box<Self> {
+        Box::new(Self {
+            message: String::from(message),
+        })
+    }
+}
+
+impl FormConstraint for Number {
+    fn validate(&self, output: &FormValue) -> bool {
+        match output {
+            FormValue::Nothing => true,
+            FormValue::Boolean(_) => true,
+            FormValue::Index(_) => true,
+            FormValue::String(value) => {
+                if let Some(chr) = value.chars().next() {
+                    if !chr.is_numeric() && chr != '-' && chr != '+' {
+                        return false;
+                    }
+                }
+                value
+                    .chars()
+                    .skip(1)
+                    .all(|x| x.is_numeric() || x == '.' || x == ',')
+                    && value.chars().filter(|&x| x == '.' || x == ',').count() <= 1
```

Comment:
And here you could do `value.parse::<f32>().is_ok()`

Reply:
I wonder is parsing to a float accepts dots as well as commas for values, also do you think keeping it that way for the same reasons as for integers is acceptable ?

Reply:
In my opinion, since it's a Rust library, validating it exactly like Rust does makes sense. Because now, if developer doesn't check the source code / documentation and tries to just `::parse<f32>` the value he can get an error even if the form is considered valid. (This also kinda applies to integer validation, but there isn't as big of a difference)
Also IIRC Rust also parses `3e10` etc., so you would also get that if you used standard Rust parsing.

Reply:
In this case parsing the input to a f64 is more relevant since floats allow values from -infinity to +infinity (aside from precision)